### PR TITLE
reuse atom signatures

### PR DESCRIPTION
**Pull request by maminrayej**
This PR tries to reuse the atom signature (sha256 hash of the atom) in the manifest to prevent recomputing the hash for that atom.
This will be done in two steps:
1. Add support for two hashing algorithms in `ModuleHash`, namely `xxhash` and `sha256`.
2. Use `sha256` for hashes of atoms in order to be able to reuse their signature.

### CODE REVIEW

File: lib/wasix/src/runtime/module_cache/types.rs [hunk 1]
```diff
@@ -127,50 +128,122 @@ impl CacheError {
...
+            let key = key.as_slice().try_into().unwrap();
+            ModuleHash::Sha256(key)
+        } else {
+            panic!("Only keys with size 8 or 32 are accepted")
+        }
+    }
+
+    /// Creates a random xxhash for the module
+    pub fn xxhash_random() -> Self {
+        Self::random::<8>()
```
Comment (syrusakbary) [inline lib/wasix/src/runtime/module_cache/types.rs:163]: This is not a proper implementation. This should return `Self::XXHash(...)`.

File: lib/wasix/src/runtime/module_cache/types.rs [hunk 2]
```diff
@@ -127,50 +128,122 @@ impl CacheError {
...
+    }
+
+    /// Creates a random sha256 hash for the module
+    pub fn sha256_random() -> Self {
+        Self::random::<32>()
```
Comment (syrusakbary) [inline lib/wasix/src/runtime/module_cache/types.rs:168]: This is not a proper implementation. This should return `Self::SHA256(...)`.

File: lib/wasix/src/runtime/module_cache/types.rs [hunk 3]
```diff
@@ -127,50 +128,122 @@ impl CacheError {
...
+        if N == 8 {
+            let key = key.as_slice().try_into().unwrap();
+            ModuleHash::XXHash(key)
+        } else if N == 32 {
+            let key = key.as_slice().try_into().unwrap();
+            ModuleHash::Sha256(key)
+        } else {
+            panic!("Only keys with size 8 or 32 are accepted")
+        }
+    }
```
Comment (syrusakbary) [inline lib/wasix/src/runtime/module_cache/types.rs:159]: This function is not proper. We send a different type depending on the length, which is highly confusion and can lead to many errors

File: lib/wasix/src/runtime/module_cache/types.rs [hunk 4]
```diff
@@ -127,50 +128,122 @@ impl CacheError {
...
+                }
+            }
+            ModuleHash::Sha256(bytes) => {
+                if N == 32 {
+                    bytes.as_slice().try_into().unwrap()
+                } else {
+                    panic!("Requested for {N} bytes, but Sha256 is 32 bytes")
+                }
+            }
+        }
```
Comment (syrusakbary) [inline lib/wasix/src/runtime/module_cache/types.rs:230]: Why we need to expose this?

File: lib/wasix/src/runtime/module_cache/types.rs [hunk 5]
```diff
@@ -127,50 +128,122 @@ impl CacheError {
...
+    }
+
+    /// Get the raw hash.
+    pub fn as_bytes<const N: usize>(self) -> [u8; N] {
+        match self {
+            ModuleHash::XXHash(bytes) => {
+                if N == 8 {
+                    bytes.as_slice().try_into().unwrap()
+                } else {
+                    panic!("Requested for {N} bytes, byt XXHash is 8 bytes");
```
Comment (syrusakbary) [inline lib/wasix/src/runtime/module_cache/types.rs:220]: I'm not a fan of panics here...

File: lib/wasix/src/syscalls/journal/mod.rs [hunk 6]
```diff
@@ -31,9 +31,9 @@ pub struct JournalSyscallPlayer<'a, 'c> {
     pub ctx: FunctionEnvMut<'c, WasiEnv>,
     pub bootstrapping: bool,
 
-    pub journal_module_hash: Option<[u8; 8]>,
+    pub journal_module_hash: Option<Vec<u8>>,
```
Comment (syrusakbary) [inline lib/wasix/src/syscalls/journal/mod.rs:34]: We can probably use a Boxed slice instead, as we'll need no conversion? (`Box<[u8]>`)

File: lib/wasix/src/syscalls/journal/mod.rs [hunk 7]
```diff
@@ -31,9 +31,9 @@ pub struct JournalSyscallPlayer<'a, 'c> {
...
     pub rewind: Option<RewindState>,
-    pub cur_module_hash: [u8; 8],
+    pub cur_module_hash: Vec<u8>,
```
Comment (syrusakbary) [inline lib/wasix/src/syscalls/journal/mod.rs:36]: We can probably use a Boxed slice instead, as we'll need no conversion? (`Box<[u8]>`)

File: lib/wasix/src/runtime/package_loader/load_package_tree.rs [hunk 8]
```diff
@@ -213,13 +217,32 @@ fn legacy_atom_hack(
...
+        .expect("malformed atom signature: does not have sha256 prefix");
+    let sha256 = base64::prelude::BASE64_STANDARD
+        .decode(base64_encoded)
+        .expect("malformed base64 encoded hash");
+    let hash: [u8; 32] = sha256
+        .as_slice()
+        .try_into()
+        .expect("sha256 hash is not 32 bytes");
+
+    ModuleHash::sha256_from_bytes(hash)
```
Comment (syrusakbary) [inline lib/wasix/src/runtime/package_loader/load_package_tree.rs:243]: I think this is quite hacky. We should just have a `FromStr` / `IntoStr` impl for the `ModuleHash`, that simply can do the `sha256:abc` -> `ModuleHash::SHA256(..)` conversion and back.

Comment (maminrayej) [reply lib/wasix/src/runtime/package_loader/load_package_tree.rs:243]: There are no hacks here. I'm just doing the steps that were taken in `pirita`, but in reverse. But I agree that your way may be more _convenient_.

File: lib/wasix/src/syscalls/journal/mod.rs (hunk 6 above)
Comment (maminrayej) [reply lib/wasix/src/syscalls/journal/mod.rs:34]: My main concern was introducing an allocation here which both `Vec` and `Box` do unfortunately. `Vec` already derefs to a slice, so I don't think here is any advantage in using `Box<[u8]>`.

File: lib/wasix/src/runtime/package_loader/load_package_tree.rs (hunk 8 above)
Comment (syrusakbary) [reply lib/wasix/src/runtime/package_loader/load_package_tree.rs]: Let's do what's more convenient, is better and we can reuse it in more places. Maybe we can even add it to wasmer-config. In any case, let's do it proper

File: lib/wasix/src/syscalls/journal/mod.rs (hunk 6 above)
Comment (syrusakbary) [reply lib/wasix/src/syscalls/journal/mod.rs:34]: Size will not be changed, so `Box<[u8]>` would probably be more optimal. I'm not religious about this, but as `Box<[u8]>` has a fixed slice, less memory will be allocated than for the vector.

This is a nit pick in any case

File: lib/wasix/src/runtime/package_loader/load_package_tree.rs (hunk 8 above)
Comment (theduke) [reply lib/wasix/src/runtime/package_loader/load_package_tree.rs]: This should not panick, but return an error instead...

**Comment by theduke** [comment]
Note: we'll want a validation function that validates all these invariants to be correct, and run that validation on webc upload in the future.

File: lib/wasix/src/syscalls/journal/mod.rs (hunk 6 above)
Comment (maminrayej) [reply lib/wasix/src/syscalls/journal/mod.rs:34]: Ah you're right. I didn't consider the size.

**Review by syrusakbary — APPROVED** [review]

File: lib/cli/src/commands/run/mod.rs [hunk 9]
```diff
@@ -717,7 +717,7 @@ impl ExecutableTarget {
                 let module = unsafe { Module::deserialize_from_file(&engine, path)? };
                 let module_hash = {
                     let wasm = std::fs::read(path)?;
-                    ModuleHash::hash(wasm)
+                    ModuleHash::xxhash(wasm)
```
Comment (syrusakbary) [inline lib/cli/src/commands/run/mod.rs:720]: I just realized that this is completely wrong, `let wasm = std::fs::read(path)?` is not a wasm file, but a precompiled artifact.
Thus, we shall not use that as the hash in any way.

Comment (syrusakbary) [reply lib/cli/src/commands/run/mod.rs:720]: Issue: https://github.com/wasmerio/wasmer/issues/4632

Comment (maminrayej) [reply lib/cli/src/commands/run/mod.rs:720]: Since we decided to track this bug in a separate issue and solve it later on, I'll resolve this comment.

**Review by syrusakbary — APPROVED** [review]

**Pull request merged**